File: server/src/main/kotlin/com/example/reconix/server/service/OcrExtractionService.kt
package com.example.reconix.server.service

import com.example.reconix.shared.OcrExtractedData
import com.example.reconix.shared.OcrLineItem
import org.apache.pdfbox.Loader
import org.apache.pdfbox.text.PDFTextStripper
import java.io.File

/**
 * OCR & Data Extraction Service
 * Uses Apache PDFBox for text-based PDF extraction
 * Falls back to simulated extraction for scanned/image PDFs
 */
class OcrExtractionService {

    /**
     * Extract data from an invoice PDF file
     * @param filePath Path to the PDF file
     * @return Extracted data with confidence score
     */
    fun extractFromPdf(filePath: String): OcrExtractedData {
        val file = File(filePath)
        if (!file.exists()) {
            return createEmptyResult("File not found: $filePath")
        }

        return try {
            val text = extractTextFromPdf(file)
            if (text.isBlank()) {
                // Scanned PDF - no text layer, would need OCR
                println("🔍 No text found in PDF - appears to be scanned image")
                createSimulatedResult()
            } else {
                parseExtractedText(text)
            }
        } catch (e: Exception) {
            println("🔍 ❌ PDF extraction error: ${e.message}")
            createEmptyResult("Extraction failed: ${e.message}")
        }
    }

    /**
     * Extract data from raw bytes (for email attachments)
     */
    fun extractFromBytes(bytes: ByteArray, fileName: String): OcrExtractedData {
        val tempFile = File.createTempFile("invoice_", ".pdf")
        try {
            tempFile.writeBytes(bytes)
            return extractFromPdf(tempFile.absolutePath)
        } finally {
            tempFile.delete()
        }
    }

    /**
     * Extract raw text from a PDF using PDFBox
     */
    private fun extractTextFromPdf(file: File): String {
        val document = Loader.loadPDF(file)
        return try {
            val stripper = PDFTextStripper()
            stripper.getText(document).trim()
        } finally {
            document.close()
        }
    }

    /**
     * Parse extracted text to find invoice data using regex patterns
     */
    private fun parseExtractedText(text: String): OcrExtractedData {
        val lines = text.lines()
        var confidence = 0.0
        var matchCount = 0
        val totalChecks = 4 // PO#, vendor, total, items

        // 1. Find PO Reference Number
        val poPattern = Regex("""(?i)(?:PO|P\.O\.|Purchase Order)[#:\s-]*(\w+-?\d+)""")
        val poMatch = poPattern.find(text)
        val detectedPoNumber = poMatch?.groupValues?.get(1)
        if (detectedPoNumber != null) matchCount++

        // 2. Find Vendor Name - look for common invoice header patterns
        val vendorPattern = Regex("""(?i)(?:from|vendor|supplier|company)[:\s]*([A-Za-z][A-Za-z\s&.,]+)""")
        val vendorMatch = vendorPattern.find(text)
        val vendorName = vendorMatch?.groupValues?.get(1)?.trim()
        if (vendorName != null) matchCount++

        // 3. Find Total Amount
        val totalPattern = Regex("""(?i)(?:total|amount due|grand total|balance due)[:\s]*\$?([\d,]+\.?\d*)""")
        val totalMatch = totalPattern.find(text)
        val totalAmount = totalMatch?.groupValues?.get(1)?.replace(",", "")?.toDoubleOrNull()
        if (totalAmount != null) matchCount++

        // 4. Extract Line Items - look for quantity × price patterns
        val lineItemPattern = Regex("""(\d+)\s+(.+?)\s+\$?([\d,]+\.?\d{0,2})\s+\$?([\d,]+\.?\d{0,2})""")
        val lineItems = lineItemPattern.findAll(text).map { match ->
            OcrLineItem(
                description = match.groupValues[2].trim(),
                quantity = match.groupValues[1].toIntOrNull(),
                unitPrice = match.groupValues[3].replace(",", "").toDoubleOrNull(),
                amount = match.groupValues[4].replace(",", "").toDoubleOrNull()
            )
        }.toList()
        if (lineItems.isNotEmpty()) matchCount++

        // Calculate confidence score
        confidence = (matchCount.toDouble() / totalChecks) * 100.0

        println("🔍 Extraction results: PO=$detectedPoNumber, Vendor=$vendorName, Total=$totalAmount, Items=${lineItems.size}, Confidence=$confidence%")

        return OcrExtractedData(
            detectedPoNumber = detectedPoNumber,
            vendorName = vendorName,
            lineItems = lineItems,
            totalAmount = totalAmount,
            confidenceScore = confidence
        )
    }

    /**
     * Create a simulated extraction result (for demo/testing when OCR is not available)
     */
    private fun createSimulatedResult(): OcrExtractedData {
        return OcrExtractedData(
            detectedPoNumber = "PO-001",
            vendorName = "Office Supplies Co.",
            lineItems = listOf(
                OcrLineItem("Printer Paper (Box)", 50, 25.00, 1250.00),
                OcrLineItem("Ink Cartridges", 10, 25.00, 250.00)
            ),
            totalAmount = 1500.00,
            confidenceScore = 65.0 // Lower confidence for simulated
        )
    }

    private fun createEmptyResult(errorMessage: String): OcrExtractedData {
        println("🔍 ⚠️ $errorMessage")
        return OcrExtractedData(
            detectedPoNumber = null,
            vendorName = null,
            lineItems = emptyList(),
            totalAmount = null,
            confidenceScore = 0.0
        )
    }
}

File: python_server/services/ocr_extraction_service.py
"""
services/ocr_extraction_service.py - OCR.space API integration + regex parsing
Uses the OCR.space REST API (key configured in core/config.py) to extract text
from PDF and image (JPG/PNG) files, then parses the text into structured data.
"""
import re
import os
import tempfile
from typing import List, Optional

import requests

from core.config import settings
from schemas import OcrExtractedData, OcrLineItem

# Supported file extensions
ALLOWED_EXTENSIONS = {".pdf", ".jpg", ".jpeg", ".png"}
# MIME types accepted by OCR.space
_MIME_MAP = {
    ".pdf":  "application/pdf",
    ".jpg":  "image/jpeg",
    ".jpeg": "image/jpeg",
    ".png":  "image/png",
}


def is_supported_file(file_name: str) -> bool:
    """Return True if the file extension is one we can process."""
    ext = os.path.splitext(file_name)[1].lower()
    return ext in ALLOWED_EXTENSIONS


def extract_from_pdf(file_path: str, file_name: str | None = None) -> OcrExtractedData:
    """
    Extract data from a file already saved to disk.
    file_name is used to determine the MIME type; falls back to basename if omitted.
    """
    if not os.path.exists(file_path):
        return _empty_result(f"File not found: {file_path}")

    name = file_name or os.path.basename(file_path)
    ext  = os.path.splitext(name)[1].lower()

    if ext not in ALLOWED_EXTENSIONS:
        return _empty_result(f"Unsupported file type: '{ext}'. Please upload a PDF, JPG, or PNG.")

    text = _call_ocr_api(file_path, name)
    if not text or not text.strip():
        return _empty_result("OCR returned no text – file may be blank or unreadable")

    # ── DEBUG: print full raw extracted text to terminal ──────────────────
    print("\n" + "═" * 60, flush=True)
    print(f"📄  OCR RAW TEXT  ─  {name}", flush=True)
    print("═" * 60, flush=True)
    print(text, flush=True)
    print("═" * 60 + "\n", flush=True)

    result = _parse_text(text)

    # ── DEBUG: print parsed structured fields to terminal ──────────────────────
    print("📊  OCR PARSED FIELDS:", flush=True)
    print(f"   Vendor      : {result.vendorName}", flush=True)
    print(f"   PO Number   : {result.detectedPoNumber}", flush=True)
    print(f"   Total Amount: {result.totalAmount}", flush=True)
    print(f"   Confidence  : {result.confidenceScore}%", flush=True)
    print(f"   Line Items  : {len(result.lineItems)} item(s)", flush=True)
    for i, li in enumerate(result.lineItems, 1):
        print(f"     [{i}] {li.description}  qty={li.quantity}  price={li.unitPrice}", flush=True)
    print(flush=True)

    return result


def extract_from_bytes(file_bytes: bytes, file_name: str) -> OcrExtractedData:
    """
    Extract data from raw bytes (e.g. from a FastAPI UploadFile).
    Returns an empty result immediately for unsupported extensions.
    """
    ext = os.path.splitext(file_name)[1].lower()
    if ext not in ALLOWED_EXTENSIONS:
        return _empty_result(
            f"Unsupported file type '{ext}'. Please upload a PDF, JPG, or PNG."
        )

    suffix = ext or ".pdf"
    with tempfile.NamedTemporaryFile(suffix=suffix, delete=False) as tmp:
        tmp.write(file_bytes)
        tmp_path = tmp.name
    try:
        return extract_from_pdf(tmp_path, file_name)
    finally:
        try:
            os.unlink(tmp_path)
        except OSError:
            pass


# ─────────────────────────────────────────────────────────────────────────────
# OCR.space API call
# ─────────────────────────────────────────────────────────────────────────────

def _call_ocr_api(file_path: str, file_name: str) -> str:
    """
    Send file to OCR.space and return the extracted plain text.
    Returns "" on any error so callers treat it as no-text.
    """
    ext      = os.path.splitext(file_name)[1].lower()
    mime     = _MIME_MAP.get(ext, "application/octet-stream")
    file_type = ext.lstrip(".")  # "pdf" | "jpg" | "jpeg" | "png"

    print(f"🔍 Sending '{file_name}' to OCR.space API (engine 2)…")
    try:
        with open(file_path, "rb") as fh:
            response = requests.post(
                settings.ocr_api_url,
                data={
                    "apikey":            settings.ocr_api_key,
                    "language":          "eng",
                    "OCREngine":         "2",        # Engine 2 – more accurate
                    "filetype":          file_type.upper(),
                    "scale":             "true",
                    "isTable":           "true",     # Structured invoice tables
                    "detectOrientation": "true",
                },
                files={"file": (file_name, fh, mime)},
                timeout=60,
            )
        response.raise_for_status()
        payload = response.json()

        if payload.get("IsErroredOnProcessing"):
            errors = payload.get("ErrorMessage") or payload.get("ErrorDetails") or "Unknown OCR error"
            print(f"🔍 ❌ OCR.space error: {errors}")
            return ""

        results = payload.get("ParsedResults") or []
        if not results:
            print("🔍 ⚠️ OCR.space returned no ParsedResults")
            return ""

        text = "\n".join(r.get("ParsedText", "") for r in results).strip()
        print(f"🔍 ✅ OCR.space extracted {len(text)} chars")
        return text

    except requests.exceptions.Timeout:
        print("🔍 ❌ OCR.space API timed out")
        return ""
    except requests.exceptions.RequestException as exc:
        print(f"🔍 ❌ OCR.space request failed: {exc}")
        return ""
    except Exception as exc:
        print(f"🔍 ❌ Unexpected OCR error: {exc}")
        return ""


# ─────────────────────────────────────────────────────────────────────────────
# Text parsing
# ─────────────────────────────────────────────────────────────────────────────

def _parse_text(text: str) -> OcrExtractedData:
    match_count = 0
    total_checks = 4

    # PO number
    po_match = re.search(r'(?i)(?:PO|P\.O\.|Purchase Order)[#:\s-]*(\w+-?\d+)', text)
    detected_po = po_match.group(1) if po_match else None
    if detected_po:
        match_count += 1

    # Vendor name
    vendor_match = re.search(r'(?i)(?:from|vendor|supplier|company)[:\s]*([A-Za-z][A-Za-z\s&.,]+)', text)
    vendor_name = vendor_match.group(1).strip() if vendor_match else None
    if vendor_name:
        match_count += 1

    # Total amount
    total_match = re.search(
        r'(?i)(?:total|amount due|grand total|balance due|invoice total)[:\s]*\$?([\d,]+\.?\d*)', text
    )
    total_amount: Optional[float] = None
    if total_match:
        try:
            total_amount = float(total_match.group(1).replace(",", ""))
            match_count += 1
        except ValueError:
            pass

    # Line items: qty  description  unit_price  total
    line_item_pattern = re.compile(r'(\d+)\s+(.+?)\s+\$?([\d,]+\.?\d{0,2})\s+\$?([\d,]+\.?\d{0,2})')
    line_items: List[OcrLineItem] = []
    for m in line_item_pattern.finditer(text):
        try:
            line_items.append(OcrLineItem(
                description=m.group(2).strip(),
                quantity=int(m.group(1)),
                unitPrice=float(m.group(3).replace(",", "")),
                amount=float(m.group(4).replace(",", "")),
            ))
        except ValueError:
            pass
    if line_items:
        match_count += 1

    confidence = (match_count / total_checks) * 100.0
    return OcrExtractedData(
        detectedPoNumber=detected_po,
        vendorName=vendor_name,
        lineItems=line_items,
        totalAmount=total_amount,
        confidenceScore=confidence,
    )


def _empty_result(error_message: str) -> OcrExtractedData:
    print(f"🔍 ⚠️ {error_message}")
    return OcrExtractedData(
        detectedPoNumber=None,
        vendorName=None,
        lineItems=[],
        totalAmount=None,
        confidenceScore=0.0,
    )
